### ISSUE TITLE
got issue with the self driving 

### ISSUE THREAD

**Issue by koi0823** [trackmania-rl/tmrl#91]
i have be try to record a reward on track 
![Screenshot 2024-04-26 185119](https://github.com/trackmania-rl/tmrl/assets/86875051/c1e790c5-f80b-4c4c-a343-b228418a98a4)
![Screenshot 2024-04-26 185200](https://github.com/trackmania-rl/tmrl/assets/86875051/cb9ff10f-e98f-4700-80d4-6bed93fda82a)

Later, when I attempted to run the check environment, all of the errors disappeared. I'm not sure whether this is a bug or what, but it kept running on the respawn location.

**Comment by yannbouteiller** [comment]
Oh dang, have I introduced a bug in 0.6.3?

The fact that it only captured two points is what causes the issue. Why did you close the issue, is this somehow fixed?

If not, are you using Windows or Linux please?

**Comment by koi0823** [comment]
I'm using a window, and I believe there is a parameter issue since if I keep running, the resolution will change to 256 X 128; I'm not sure why.

**Comment by koi0823** [comment]
i
![Screenshot 2024-04-27 024656](https://github.com/trackmania-rl/tmrl/assets/86875051/7b47427b-687b-4e87-9ca8-48411e7f64bd)
 did fix the height and width

**Comment by koi0823** [comment]
![Uploading Screenshot 2024-04-27 024909.png…]()

**Comment by koi0823** [comment]
![Uploading Screenshot 2024-04-27 024909.png…]()

**Comment by koi0823** [comment]
![Screenshot 2024-04-27 024909](https://github.com/trackmania-rl/tmrl/assets/86875051/0a6db545-0b42-41b0-b569-20cea3b5c4af)
so this is my version yes is 0.6.3

**Comment by koi0823** [comment]
after that i found out find in CMD is useless then terminal in VSC firstly i try run in VSC

**Comment by yannbouteiller** [comment]
I tested `0.6.3` for both `--record-reward` and `--check-environment` on Windows and both worked as intended on my machine.

Your output for `--record-reward` is extremely strange, it looks like the car instantly teleported to the finish line. The "number of initial position" should never be 2.

**Comment by koi0823** [comment]
I'm utilizing the entire environment to run it, so is it because the road isn't balanced that I can't record the award and check the environment?

**Comment by koi0823** [comment]
maybe is the dent of the road that why the lidar cannot scan ?

**Comment by yannbouteiller** [comment]
No you should be fine,, this has nothing to do with the lidar. I recorded a reward on the 1st Spring 2024 track myself and it worked properly.

Something seems wrong with your OpenPlanet installation, as if you were getting receiving only 2 positions while recording the reward for some reason.

Maybe some external program is sending trash on port 9000?

To record the reward, you are supposed to set the car at the beginning of the track, press e, and then drive to the finish line. When you cross the finish line, you should see a message telling you that it has captured a few thousand initial positions (your problem is that it only captures 2 positions for some weird reason).

**Comment by koi0823** [comment]
yaya but i have to run at cmd

**Comment by koi0823** [comment]
![Screenshot 2024-04-28 192936](https://github.com/trackmania-rl/tmrl/assets/86875051/133a8bd3-ed7a-40e9-b6d0-c088599ea5fd)
cus i ask what is the loss_actor is it become worst and worst cus i dont know is it training better and better

**Comment by koi0823** [comment]
INFO:root:=== epoch 42/10000 = round 76/100 ====================================
INFO:root:memory_len                  293223
round_time                2.991769
idle_time                      0.0
loss_actor                -0.67442
loss_critic               0.524971
return_test                    0.0
return_train                 47.25
episode_length_test            0.0
episode_length_train         127.0
sampling_duration         0.006929
training_step_duration    0.008019

**Comment by yannbouteiller** [comment]
LIDAR is faster and easier to train, but it only works on plain grey road with black borders. Full is more general and works on any track. It is however harder to find good hyperparameters with Full, and training takes a long time / requires a high-end GPU.

Everything is located in the `TmrlData` folder, indluding the reward. However, this is a pickle file, so exploring this file will not help you debugging, unless you unpickle and explore its content in a pickle script. But from your logs, what you would see if you were doing that would probably be a bunch of 3D points forming a straight line between your 2 initial captured positions.

**Comment by koi0823** [comment]
Regarding the third column you said, I believe it should be alright as I ran it well and deleted it once beforehand. I used VSC to execute the Python -m tmrl --test works good, but the reward record appears to be stuck—possibly because I kept pushing E—so it isn't recording. 

in other way i use cmd to test the python -m tmrl --record-reward is all fine

**Comment by koi0823** [comment]
![Uploading Screenshot 2024-04-29 002943.png…]()

**Comment by koi0823** [comment]
![Screenshot 2024-04-29 002943](https://github.com/trackmania-rl/tmrl/assets/86875051/d7928a8d-d687-4d8c-ac1c-0e900de381e8)

idk is it okie or not but is training

**Comment by yannbouteiller** [comment]
Oh yes, you don't want to keep pushi.g e, just press it once at the start of the track, then drive normally to the end, and when you cross the finish line the script should automatically compute the reward function from positions it captured in between.

Are you using the Full environment? If you are, it should automatically rescale the trackmania window to something smaller (unless you manually changed the corresponding parameters in `config.json`).

**Comment by koi0823** [comment]
![image](https://github.com/trackmania-rl/tmrl/assets/86875051/cc8ba887-93dc-4541-8fc0-7294c0fc1077)

**Comment by koi0823** [comment]
![image](https://github.com/trackmania-rl/tmrl/assets/86875051/8fdc0ff7-7d1f-447f-98d5-2a0c1ea7f181)

**Comment by koi0823** [comment]
File "C:\Users\user\AppData\Local\Programs\Python\Python310\lib\site-packages\gymnasium\core.py", line 414, in reset
    return self.env.reset(seed=seed, options=options)
  File "C:\Users\user\AppData\Local\Programs\Python\Python310\lib\site-packages\gymnasium\wrappers\order_enforcing.py", line 61, in reset
    return self.env.reset(**kwargs)
  File "C:\Users\user\AppData\Local\Programs\Python\Python310\lib\site-packages\rtgym\envs\real_time_env.py", line 514, in reset
    elt, info = self.interface.reset(seed=seed, options=options)
  File "C:\Users\user\AppData\Local\Programs\Python\Python310\lib\site-packages\tmrl\custom\custom_gym_interfaces.py", line 147, in reset
    self.reset_common()
  File "C:\Users\user\AppData\Local\Programs\Python\Python310\lib\site-packages\tmrl\custom\custom_gym_interfaces.py", line 137, in reset_common
    self.initialize()
  File "C:\Users\user\AppData\Local\Programs\Python\Python310\lib\site-packages\tmrl\custom\custom_gym_interfaces.py", line 89, in initialize
    self.initialize_common()
  File "C:\Users\user\AppData\Local\Programs\Python\Python310\lib\site-packages\tmrl\custom\custom_gym_interfaces.py", line 73, in initialize_common
    self.j = vg.VX360Gamepad()
  File "C:\Users\user\AppData\Local\Programs\Python\Python310\lib\site-packages\vgamepad\win\virtual_gamepad.py", line 115, in __init__
    super().__init__()
  File "C:\Users\user\AppData\Local\Programs\Python\Python310\lib\site-packages\vgamepad\win\virtual_gamepad.py", line 60, in __init__
    assert vcli.vigem_target_is_attached(self._devicep), "The virtual device could not connect to ViGEmBus."

**Comment by koi0823** [comment]
Why am I suddenly unable to train?

**Comment by koi0823** [comment]
hey i have a question for the time_step_timeout_factor cus i set it 80 but it will stop when is 48 second 

![image](https://github.com/trackmania-rl/tmrl/assets/86875051/1c51c230-25dc-45c3-8c1c-be738f04ee31)

**Comment by koi0823** [comment]
![image](https://github.com/trackmania-rl/tmrl/assets/86875051/3023dd4f-02b2-4194-8a65-ba7e145b86dd)

**Comment by yannbouteiller** [comment]
Wow, don't do this. The timestep timeout factor should not exceed 2.0, otherwise it becomes entirely meaningless.

Closing as this is not a tmrl issue. Please open a thread in the discussions section for help and questions.

**Comment by koi0823** [comment]
> Wow, don't do this. The timestep timeout factor should not exceed 2.0, otherwise it becomes entirely meaningless.
> 
> Closing as this is not a tmrl issue. Please open a thread in the discussions section for help and questions.

u mean the timestep timeout factor should not exceed 2.0 
 keep respawn and respawn so i chatgpt it and is keep this ans

**Comment by koi0823** [comment]
@yannbouteiller  then how should i set my time and second cus is weird keep respawn then didnt stop

**Comment by yannbouteiller** [comment]
Respawns are not related to the timeout factor, they happen because the agent is failing to collect reward.

It probably fails to collect reward because your environment has an issue. You need to use `python -m tmrl --check-environment` to find out what.

**Comment by koi0823** [comment]
@yannbouteiller thx bro i will take a look

**Comment by koi0823** [comment]
https://youtu.be/2NkFNORkdD0?si=S3krB2NBnP2FN1P2

@yannbouteiller check this out i think my environment is fine but idk why is keep stuck at 50 second

**Comment by koi0823** [comment]
![image](https://github.com/trackmania-rl/tmrl/assets/86875051/f26b28c1-1d31-4393-8056-a789ac67b879)

**Comment by koi0823** [comment]
i didnt change anything on the config 
![image](https://github.com/trackmania-rl/tmrl/assets/86875051/ddb06790-7081-476f-88bc-5dc2ff5d6483)

**Comment by yannbouteiller** [comment]
> https://youtu.be/2NkFNORkdD0?si=S3krB2NBnP2FN1P2 video on 2.10 time is the place i located and is 50 second
> 
> ![image](https://private-user-images.githubusercontent.com/86875051/326903936-19af784c-d48d-472d-88f4-00c918a6b1d2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQ0OTU5MDYsIm5iZiI6MTcxNDQ5NTYwNiwicGF0aCI6Ii84Njg3NTA1MS8zMjY5MDM5MzYtMTlhZjc4NGMtZDQ4ZC00NzJkLTg4ZjQtMDBjOTE4YTZiMWQyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDMwVDE2NDY0NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTEyYzJiNDJiZjI5ODJiYWM4MWVkNWUxMzA3MWFkZTRjZjQ5NjM2ZDE2NmMxNmI1Yjk0NWUwODMxMmZlZmVhMTAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.yhtZu25EV6M4NqZ2VBKMClfeFjSEmqZWb2KoEo1c4Ms)
> 
> @yannbouteiller check this out i think my environment is fine but idk why is keep stuck at 50 second

The 50 seconds limit is expected, it is the default time-limit in the example TrackMania pipeline. Your environment looks sane from your screenshots.

**Comment by yannbouteiller** [comment]
You can change it in `config.json` by changing the values of both the "ep_max_len"  and the "RW_MAX_SAMPLES_PER_EPISODE" entries.

**Comment by koi0823** [comment]
It's good that the time has changed, but how can I make the parameter more pleasant?
because it continues to tremble

**Comment by yannbouteiller** [comment]
The jitter is inherent to how Soft Actor-Critic trains policies. If you want to get rid of it, one solution is to penalize large changes in the steering in the reward function. This would require you to learn Python programming, clone the repository and adapt the environment' code.

Otherwise you can play with the SAC hyperparameters in `config.json`, in particular the "alpha" term which is responsible for injecting entropy in the policy.

**Comment by koi0823** [comment]
mean i need increase the value ? to like 0.05 or like less?

![image](https://github.com/trackmania-rl/tmrl/assets/86875051/d93d7908-c9c0-4a7f-b0c9-3f34b25fe93b)

**Comment by yannbouteiller** [comment]
If you want to see less jitter, you should decrease it I believe, but this will also harm exploration.

**Comment by koi0823** [comment]
can i ask what is loss_actor ?
![image](https://github.com/trackmania-rl/tmrl/assets/86875051/c6371d39-41e4-4c4b-98a8-83c4cfb93ec0)

**Comment by yannbouteiller** [comment]
You need to read the Soft Actor-Critic paper to understand this.
